### PR TITLE
Support arrow readers for strings with DELTA_BYTE_ARRAY encoding

### LINKED ISSUE #708: 'Encoding DELTA_BYTE_ARRAY is not supported' with parquet arrow readers
**Describe the bug**
Reading parquet files with byte arrays encoded using `DELTA_BYTE_ARRAY` causes the error mentioned from the title.
Corresponding fields must be of `UTF8` converted type.

**To Reproduce**
1. read parquet file with strings encoded as `DELTA_BYTE_ARRAY` with parquet's arrow readers.

**Expected behavior**
1. file reads with no error

Actual: the aforementioned error is produced.

**Additional context**
We fixed this in our fork, I'll send a PR for review and further discussion soon.

### PR DESCRIPTION
**Pull request by ilya-biryukov**
# Which issue does this PR close?

Closes #708 .

# Rationale for this change
 
This allow to read parquet files that use the specified encoding.

# What changes are included in this PR?

We wrap existing code to read the  into `ValueDecoder`.
Also added some machinery to write and read parquet files from memory. This seems generally useful, but happy to remove it from this PR if that makes sense.

# Are there any user-facing changes?

No, it's a bugfix.

### CODE REVIEW

File: parquet/src/arrow/arrow_array_reader.rs [hunk 1]
```diff
@@ -1559,4 +1605,120 @@ mod tests {
...
+                .build(),
+        );
+        // Write a few strings.
+        let mut w = SerializedFileWriter::new(data.clone(), schema, p).unwrap();
+        let mut rg = w.next_row_group().unwrap();
+        let mut c = rg.next_column().unwrap().unwrap();
+        match &mut c {
+            ColumnWriter::ByteArrayColumnWriter(c) => {
+                c.write_batch(
+                    &[ByteArray::from("foo"), ByteArray::from("bar")],
```
Comment (alamb) [inline parquet/src/arrow/arrow_array_reader.rs:1688]: I recommend also adding a test here for `Null` / `None` (aka definition level 0)

**Review by alamb — COMMENTED** [review]
Thank you @ilya-biryukov  ! I plan to review this more carefully tomorrow morning.

cc @nevi-me

File: parquet/src/arrow/arrow_array_reader.rs [hunk 2]
```diff
@@ -1074,6 +1079,39 @@ impl ValueDecoder for VariableLenDictionaryDecoder {
...
+impl ValueDecoder for DeltaByteArrayValueDecoder {
+    fn read_value_bytes(
+        &mut self,
+        mut num_values: usize,
+        read_bytes: &mut dyn FnMut(&[u8], usize),
+    ) -> Result<usize> {
+        num_values = std::cmp::min(num_values, self.decoder.values_left());
+        let mut values_read = 0;
+        while values_read < num_values {
+            let mut buf = [ByteArray::new()];
```
Comment (alamb) [inline parquet/src/arrow/arrow_array_reader.rs:1103]: Two performance questions: 
1. I wonder if allocating a new `ByteArray` on each iteration of the loop  (aka each value read from the file) will be  overly slow. What about allocating `buf` outside of the loop?
2. Why only read one value at a time (aka why use `buf` with a size of `1`? Why not some other value (100?))

File: parquet/src/arrow/arrow_array_reader.rs [hunk 3]
```diff
@@ -1074,6 +1079,39 @@ impl ValueDecoder for VariableLenDictionaryDecoder {
...
+            let num_read = self.decoder.get(&mut buf)?;
+            assert_eq!(num_read, 1);
```
Comment (alamb) [inline parquet/src/arrow/arrow_array_reader.rs:1105]: Perhaps `debug_assert_eq` might be more appropriate here so as not to do the check in release builds?

**Review by alamb — COMMENTED** [review]
I went through this carefully @ilya-biryukov  -- nice work 👍 

I do think a test that this handles Nulls correctly is important prior to merge. My other comments are related to potential performance optimizations, which I think would be fine to do as a follow on PR if/when profiling shows it is slow.

File: parquet/src/arrow/arrow_array_reader.rs (hunk 2 above)
Comment (ilya-biryukov) [reply parquet/src/arrow/arrow_array_reader.rs]: 1. I initially thought it should not make a difference as the allocation is static. But your comment got me thinking and I did some [benchmarking](https://gist.github.com/ilya-biryukov/d053433923ec7ab9b94e897a649a92c5). Putting the let statement outside the loop is actually > 3x faster in the benchmarks, so it's definitely a good habit to do this.
2. It's a restriction that `StringArrayConverter` imposes, see https://github.com/apache/arrow-rs/blob/f438f72bb1b2e61c095c4ea580e6abe2d5048360/parquet/src/arrow/arrow_array_reader.rs#L1178
I was not eager to change it here as this focuses on fixing the bug, rather than perf optimization.

File: parquet/src/arrow/arrow_array_reader.rs (hunk 3 above)
Comment (ilya-biryukov) [reply parquet/src/arrow/arrow_array_reader.rs]: Thx, good point!

File: parquet/src/arrow/arrow_array_reader.rs (hunk 1 above)
Comment (ilya-biryukov) [reply parquet/src/arrow/arrow_array_reader.rs:1688]: Done.

**Comment by ilya-biryukov** [comment]
Thanks for the quick response! Here's an updated version.

File: parquet/src/arrow/arrow_array_reader.rs [hunk 4]
```diff
@@ -1559,4 +1605,120 @@ mod tests {
...
+        assert_eq!(batch.columns().len(), 1);
+
+        let strings = batch
+            .column(0)
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .unwrap();
+        assert_eq!(
+            strings.into_iter().collect::<Vec<_>>(),
+            vec![None, Some("foo"), None, None, Some("bar"), None]
```
Comment (alamb) [inline parquet/src/arrow/arrow_array_reader.rs:1721]: 🏅  for Nulls at start and end and middle

**Review by alamb — APPROVED** [review]
Looks great to me @ilya-biryukov  -- thank you!
